docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+index_part.json written by the last node to do writes on this tenant, whose exact key is unknown
+as we do not know what node/generation last wrote it. This is pretty simple to solve:
+
+- we can issue a ListObjectsv2 request with the index_part.json as the prefix, and load the object with
+  the highest attachment generation number, and then as soon as we've written out an index_part.json with
+  our generation suffix, we may erase the old index_part.json files.
+- This listing can be avoided with some optimizations, depending on the HA scheme in use. If HA is driven
+  entirely by tenant attachment changes, then writer nodes may write out a tiny object one time at startup,
+  where this "link" object just has the attachment generation in its name, so that future nodes can subtract
+  one from their own attachment generation to get an almost-always-right GET path for the object that will
+  lead them to the most recent writer's latest metadata.
+- Alternatively, whatever drives failover (perhaps the control plane) could also provide the hint to new
+  attachments of what the most recent active attachment was, so that the newly attached node already
+  knows the generation numbers of the node that last wrote metadata for this tenant.
+
+### Control Plane Changes
+
+#### Managing generation numbers
+
+The control plane is responsible for issuing generation numbers: this work is done in the control plane
+because without adding a consensus system to the pageservers, they do not have a way to reliably
+allocate unique generation numbers while tolerating network partitions.
+
+Assigning generation numbers to attachments is very simple, because designating a particular pageserver
+as the attachment point for a tenant is already a database transaction: this transaction just needs
+to increment the generation number as well as setting the pageserver.
+
+Assigning generation numbers to nodes at startup is a larger change: pageserver startup must be
+gated on communication from the control plane. We may either implement this via
+
+- pull: the pageserver calling up to the control plane to announce itself and request a generation number
+- push: in the opposite direction by having the control plane call down to the pageserver to give it
+  a generation number.
+
+The choice of pull or push does not affect the proposed storage format changes.
+
+#### Synchronizing attachments on pageserver startup
+
+In "pageserver startup changes" above, it is noted that the pageserver should synchronize its list of
+attachments on startup, to avoid starting redundant writes to S3 for tenants that are already being
+written from some other more up-to-date node.
+
+However, it is important to avoid O(tenant_count) communications where possible, to avoid excessive
+load on the control plane. Especially, there are two common special cases that can be made very
+fast:
+
+- A pageserver that restarts without any attachments being changed: this would happen if a pageserver
```

Comment:
If we restrict pageserver<->control plane communications to only active tenants (or even active tenants that want to delete something) then number of requests on start will be similar to number of request during the normal work.

Reply:
Revised this text.  Having changed the definition of deletion to require that pageservers validate their attachment generations before doing deletions, this whole section becomes purely an optimization.  We may let a pageserver with stale attachments proceed with ingestion and even serve reads, and they will be implicitly prevented from doing deletions by the generation number check on deletion.

We have a few options here:
1. Skip the optimization entirely: stale attachments keep ingesting and writing to S3 until the control plane gets around to detaching them.
2. Do an exhaustive check of all attachments on startup, this would have a similar CP cost to all the deletion checks we would usually spread over some ~10min timespan, so would be quite spiky.  In hand-waving terms this feels fine for ~10k tenants but probably not for ~100k tenants.
3. Do the extra optimization for the "all ok" and "none ok" cases that correspond to a fast restart and a failed-evacuated node respectively.

It's true that number 3 is quite special-casey -- it may be that we start out just doing option 1: if the control plane's reconciliation loop is running promptly, it will detach those stale tenants very soon after the pageserver starts up.  We could also implement a "grace period" on the pageserver where it won't do any S3 writes until some time threshold has passed, during which we would expect any stale ones to detach, although I prefer to avoid walltime-based behaviors where possible.

Reply:
This is the "Synchronizing attachments on pageserver startup" section in 896bfdb2299d2a4f4b1b942a35f5704ce801d020

On reflection, I think it makes sense to let the pageserver do its own validation of attachments using the same API  it consumes for the barrier on deletions.  On startup we may issue a series of batched RPCs to the control plane to avoid one mega all-attachments message.  These are read-only operations for the control plane so should be very fast.  That will be more robust than having to ensure that all control plane code properly manages some "dirty" flag on pageservers to indicate whether they require a sync.  3bc4110913a530ecfe318fe351bbfad1ad62cc2b